// remix-ide/.history/src/lib/panels-resize_20190819095436.js
const yo = require('yo-yo')
const csjs = require('csjs-inject')

const css = csjs`
  .dragbar            {
    width             : 4px;
    height            : 100%;
    cursor            : col-resize;
    z-index           : 999;
  }
  .ghostbar           {
    width             : 3px;
    background-color  : var(--primary);
    opacity           : 0.5;
    position          : absolute;
    cursor            : col-resize;
    z-index           : 9999;
    top               : 0;
    bottom            : 0;
  }
`

export default class PanelsResize {
  constructor (panel) {
    this.panel = panel
    let self = this;
    window.addEventListener('resize', function (event) {
      self.setPosition(event)
    })

    //this.setPosition(this.minWidth)
  }

  render () {
    this.minWidth = this.panel.clientWidth
    this.ghostbar = yo`<div class=${css.ghostbar}></div>`

    let mousedown = (event) => {
      event.preventDefault()
      if (event.which === 1) {
        moveGhostbar(event)
        document.body.appendChild(this.ghostbar)
        document.addEventListener('mousemove', moveGhostbar)
        document.addEventListener('mouseup', removeGhostbar)
        document.addEventListener('keydown', cancelGhostbar)
      }
    }
    
    let cancelGhostbar = (event) => {
      if (event.keyCode === 27) {
        document.body.removeChild(this.ghostbar)
        document.removeEventListener('mousemove', moveGhostbar)
        document.removeEventListener('mouseup', removeGhostbar)
        document.removeEventListener('keydown', cancelGhostbar)
      }
    }

    let moveGhostbar = (event) => { // @NOTE VERTICAL ghostbar
      //let panelWidth = calculatePanelWidth(event)
      //if (panelWidth === 0) this.hidePanel()
      //if (panelWidth <= this.panel.offsetWidth) return //todo check
      this.ghostbar.style.left = event.x + 'px'
      if (calculatePanelWidth(event) === 0 && event.x < this.panel.offsetLeft) this.showPanel()
    }

    let removeGhostbar = (event) => {
      document.body.removeChild(this.ghostbar)
      document.removeEventListener('mousemove', moveGhostbar)
      document.removeEventListener('mouseup', removeGhostbar)
      document.removeEventListener('keydown', cancelGhostbar)
      this.setPosition(event)
    }

    return yo`<div onmousedown=${mousedown} class=${css.dragbar}></div>`
  }

  calculatePanelWidth (event) {
    let panelWidth = event.x - this.panel.offsetLeft
    if (panelWidth < this.minWidth - 10) {
      return 0;
    }
    return panelWidth
  }

  setPosition (event) {
    let panelWidth = this.calculatePanelWidth(event)
    if (panelWidth !== 0) this.panel.style.width = panelWidth + 'px'
    else this.hidePanel()
  }

  hidePanel () {
    this.panel.style.display = 'none'
  }

  showPanel () {
    this.panel.style.display = 'flex'
  }
}
