// remix-ide/.history/assets/js/editor/darkTheme_20200612163603.js
/* eslint-disable */
ace.define("ace/theme/chaosRemix",["require","exports","module","ace/lib/dom"], function(acequire, exports, module) {

  exports.isDark = true;
  exports.cssClass = "ace-chaosRemix";
  exports.cssText = ".ace-chaosRemix .ace_gutter {\
  background: #2a2c3f;\
  color: #8789a1;\
  border-right: 1px solid #282828;\
  }\
  .ace-chaosRemix .ace_gutter-cell.ace_warning {\
  background-image: none;\
  background: #FC0;\
  border-left: none;\
  padding-left: 0;\
  color: #000;\
  }\
  .ace-chaosRemix .ace_gutter-cell.ace_error {\
  background-position: -6px center;\
  background-image: none;\
  background: #F10;\
  border-left: none;\
  padding-left: 0;\
  color: #000;\
  }\
  .ace-chaosRemix .ace_print-margin {\
  border-left: 1px solid #555;\
  right: 0;\
  background: #1D1D1D;\
  }\
  .ace-chaosRemix {\
  background-color: #222336;\
  color: #a2a3bd;\
  }\
  .ace-chaosRemix .ace_cursor {\
  border-left: 2px solid #FFFFFF;\
  }\
  .ace-chaosRemix .ace_cursor.ace_overwrite {\
  border-left: 0px;\
  border-bottom: 1px solid #FFFFFF;\
  }\
  .ace-chaosRemix .ace_marker-layer .ace_selection {\
  background: #494836;\
  }\
  .ace-chaosRemix .ace_marker-layer .ace_step {\
  background: rgb(198, 219, 174);\
  }\
  .ace-chaosRemix .ace_marker-layer .ace_bracket {\
  margin: -1px 0 0 -1px;\
  border: 1px solid #FCE94F;\
  }\
  .ace-chaosRemix .ace_marker-layer .ace_active-line {\
  background: #363950;\
  }\
  .ace-chaosRemix .ace_gutter-active-line {\
  background-color: #363950;\
  }\
  .ace-chaosRemix .ace_invisible {\
  color: #404040;\
  }\
  .ace-chaosRemix .ace_rparen {\
    color: #d4d7ed;\
  }\
  .ace-chaosRemix .ace_lparen {\
    color: #d4d7ed;\
  }\
  .ace-chaosRemix .ace_keyword {\
  color:#ffa76d;\
  }\
  .ace-chaosRemix .ace_keyword.ace_operator {\
  color:#eceeff;\
  }\
  .ace-chaosRemix .ace_constant {\
  color:#1EDAFB;\
  }\
  .ace-chaosRemix .ace_constant.ace_language {\
  color:#FDC251;\
  }\
  .ace-chaosRemix .ace_constant.ace_library {\
  color:#8DFF0A;\
  }\
  .ace-chaosRemix .ace_constant.ace_numeric {\
  color:#eceeff;\
  }\
  .ace-chaosRemix .ace_invalid {\
  color:#FFFFFF;\
  background-color:#990000;\
  }\
  .ace-chaosRemix .ace_invalid.ace_deprecated {\
  color:#FFFFFF;\
  background-color:#990000;\
  }\
  .ace-chaosRemix .ace_support {\
  color: #999;\
  }\
  .ace-chaosRemix .ace_support.ace_function {\
  color:#3fe2a7;\
  }\
  .ace-chaosRemix .ace_function {\
  color:#3fe2a7;\
  }\
  .ace-chaosRemix .ace_string {\
  color:#eceeff;\
  }\
  .ace-chaosRemix .ace_comment {\
  color:#a7a7a7;\
  font-style:italic;\
  padding-bottom: 0px;\
  }\
  .ace-chaosRemix .ace_type {\
  color:#75ceef;\
  }]\
  .ace-chaosRemix .ace_visibility (\
    color:#f7d777;\
  )\
  .ace-chaosRemix .ace_identifier {\
    color:#bec1dd;\
  }\
  .ace-chaosRemix .ace_modifier {\
    color:#efff2f;\
  }\
  .ace-chaosRemix .ace-boolean {\
    color:#ff86ac;\
  }\
  .ace-chaosRemix .ace_statemutability {\
    color:#FFCC00;\
  }\
  .ace-chaosRemix .ace_variable {\
  color:#e0bb83;\
  }\
  .ace-chaosRemix .ace_meta.ace_tag {\
  color:#BE53E6;\
  }\
  .ace-chaosRemix .ace_entity.ace_other.ace_attribute-name {\
  color:#4aa8fd;\
  }\
  .ace-chaosRemix .ace_markup.ace_underline {\
  text-decoration: underline;\
  }\
  .ace-chaosRemix .ace_fold-widget {\
  text-align: center;\
  }\
  .ace-chaosRemix .ace_fold-widget:hover {\
  color: #777;\
  }\
  .ace-chaosRemix .ace_fold-widget.ace_start,\
  .ace-chaosRemix .ace_fold-widget.ace_end,\
  .ace-chaosRemix .ace_fold-widget.ace_closed{\
  background: none;\
  border: none;\
  box-shadow: none;\
  }\
  .ace-chaosRemix .ace_fold-widget.ace_start:after {\
  content: '▾'\
  }\
  .ace-chaosRemix .ace_fold-widget.ace_end:after {\
  content: '▴'\
  }\
  .ace-chaosRemix .ace_fold-widget.ace_closed:after {\
  content: '‣'\
  }\
  .ace-chaosRemix .ace_indent-guide {\
  border-right:1px dotted #333;\
  margin-right:-1px;\
  }\
  .ace-chaosRemix .ace_fold { \
  background: #222; \
  border-radius: 3px; \
  color: #7AF; \
  border: none; \
  }\
  .ace-chaosRemix .ace_fold:hover {\
  background: #CCC; \
  color: #000;\
  }\
  ";
  
  var dom = acequire("../lib/dom");
  dom.importCssString(exports.cssText, exports.cssClass);
  
  });
