// remix-ide/.history/src/app/ui/landing-page/workspace_20200217172319.js
export class Workspace {
  constructor (title, description, isMain, activate, deactivate) {
    this.title = title
    this.description = description
    this.isMain = isMain
    this.activate = activate
    this.deactivate = deactivate
  }
}

export const defaultWorkspaces = (appManager) => {
  return [
    new Workspace(
      'Solidity',
      'Writing smart contracts. It is used for implementing smart contracts on various blockchain platforms',
      true,
      () => {
        appManager.activatePlugin('solidity')
        appManager.activatePlugin('udapp')
        appManager.activatePlugin('solidityStaticAnalysis')
        appManager.activatePlugin('solidityUnitTesting')
      }, () => {}),
    new Workspace(
      'Vyper',
      'Vyper is a contract-oriented, pythonic programming language that targets the Ethereum Virtual Machine (EVM)',
      true,
      () => {
        appManager.activatePlugin('vyper')
        appManager.activatePlugin('udapp')
      }, () => {}),
    new Workspace('Debugger', 'Debug transactions with remix', false, () => {
      appManager.activatePlugin('debugger')
    }, () => {}),
    new Workspace('Pipeline', '', false, () => {
      appManager.activatePlugin('solidity')
      appManager.activatePlugin('pipeline')
      appManager.activatePlugin('udapp')
    })
  ]
}
