// remix-ide/.history/src/app/files/filePovider_20190916145229.js
'use strict'

var EventManager = require('../../lib/events')

class FileProvider {
  constructor (name, isReadOnly, storage) {
    this.event = new EventManager()
    this.storage = storage
    this.type = name
    this.structFile = '.' + name + '.tree'
    this.tree = {}
    this.isReadOnly = isReadOnly

    this.files = {}
    this.paths = {}
    this.normalizedNames = {} // contains the raw url associated with the displayed path
    this.paths[type] = {}
  }



  }

  exists (path, cb) {
    cb(null, this._exists(path))
  }

  updateRefs (path, type) {
    var split = path.split('/') // this should be unprefixed path
    var crawlpath = this.tree
    var intermediatePath = ''
    split.forEach((pathPart, index) => {
      intermediatePath += pathPart
      if (!crawlpath[pathPart]) crawlpath[intermediatePath] = {}
      if (index < split.length - 1) {
        crawlpath = crawlpath[intermediatePath]
        intermediatePath += '/'
      } else if (type === 'add') {
        crawlpath[intermediatePath] = path
      } else if (type === 'remove' && crawlpath[intermediatePath]) {
        delete crawlpath[intermediatePath]
      }
    })
    this.storage.set(this.structFile, JSON.stringify(this.tree))
  }

  _exists (path) {
    var unprefixedpath = this.removePrefix(path)
    return this.storage.exists(unprefixedpath)
  }

  init (cb) {
    var tree = this.storage.get(this.structFile)
    this.tree = tree ? JSON.parse(tree) : {}
    if (cb) cb()
  }

  get (path, cb) {
    var unprefixedpath = this.removePrefix(path)
    var content = this.storage.get(unprefixedpath)
    if (cb) {
      cb(null, content)
    }
    return content
  }

  set (path, content, cb) {
    var unprefixedpath = this.removePrefix(path)
    this.updateRefs(unprefixedpath, 'add')
    var exists = this.storage.exists(unprefixedpath)
    if (!this.storage.set(unprefixedpath, content)) {
      if (cb) cb('error updating ' + path)
      return false
    }
    if (!exists) {
      this.event.trigger('fileAdded', [this.type + '/' + unprefixedpath, false])
    } else {
      this.event.trigger('fileChanged', [this.type + '/' + unprefixedpath])
    }
    if (cb) cb()
    return true
  }

  addReadOnly (path, content, rawPath) {
    path = this.removePrefix(path)
    try { // lazy try to format JSON
      content = JSON.stringify(JSON.parse(content), null, '\t')
    } catch (e) {}
    if (!rawPath) rawPath = path
    // splitting off the path in a tree structure, the json tree is used in `resolveDirectory`
    var split = path
    var folder = false
    while (split.lastIndexOf('/') !== -1) {
      var subitem = split.substring(split.lastIndexOf('/'))
      split = split.substring(0, split.lastIndexOf('/'))
      if (!this.paths[this.type + '/' + split]) {
        this.paths[this.type + '/' + split] = {}
      }
      this.paths[this.type + '/' + split][split + subitem] = { isDirectory: folder }
      folder = true
    }
    this.paths[this.type][split] = { isDirectory: folder }
    this.files[path] = content
    this.normalizedNames[rawPath] = path
    this.event.trigger('fileAdded', [this.type + '/' + path, true])
    return true
  }

  isReadOnly (path) {
    return this.isReadOnly
  }

  remove (path) {
    var unprefixedpath = this.removePrefix(path)
    this.updateRefs(unprefixedpath, 'remove')
    if (!this._exists(unprefixedpath)) {
      return false
    }
    if (!this.storage.remove(unprefixedpath)) {
      return false
    }
    this.event.trigger('fileRemoved', [this.type + '/' + unprefixedpath])
    return true
  }

  rename (oldPath, newPath, isFolder) {
    var unprefixedoldPath = this.removePrefix(oldPath)
    var unprefixednewPath = this.removePrefix(newPath)
    this.updateRefs(unprefixedoldPath, 'remove')
    this.updateRefs(unprefixednewPath, 'add')
    if (this.storage.exists(unprefixedoldPath)) {
      if (!this.storage.rename(unprefixedoldPath, unprefixednewPath)) {
        return false
      }
      this.event.trigger('fileRenamed', [
        this.type + '/' + unprefixedoldPath,
        this.type + '/' + unprefixednewPath,
        isFolder
      ])
      return true
    }
    return false
  }

  list () {
    return this.files
  }

  resolveDirectory (path, callback) {
    if (path[0] === '/') path = path.substring(1)
    if (!path) return callback(null, { [this.type]: {} })
    var tree = {}
    path = this.removePrefix(path)

    var split = path.split('/') // this should be unprefixed path
    var crawlpath = this.tree
    split.forEach((pathPart, index) => {
      if (crawlpath[pathPart]) crawlpath = crawlpath[pathPart]
    })

    for (var item in crawlpath) {
      tree[item] = { isDirectory: typeof crawlpath[item] !== 'string' }
    }
    callback(null, tree)
  }

  removePrefix (path) {
    path = path.indexOf(this.type) === 0 ? path.replace(this.type, '') : path
    if (path[0] === '/') return path.substring(1)
    return path
  }
}

module.exports = FileProvider
